fix the abnormal GPU memory occupation of multimodal model continued to increase until OOM

huangtingwei9988 opened the pull request:
## Motivation
We used sglang to deploy the Qwen2.5-vl-72b （8*H20）. OOM occurred after running for about an hour. We are very confused because` --mem-fraction-static` is only 0.80. As shown in the figure below, when the deployment is just completed, there is about 20G of memory available on each GPU.
<img width="585" alt="image" src="https://github.com/user-attachments/assets/e3b3c726-5026-4d24-b0d8-21ace1b38b60" />
However, the memory usage continued to increase. After half an hour, each GPU had less than 7G, and finally OOM occurred.
![image](https://github.com/user-attachments/assets/95271534-70d9-437b-b97f-7439251ddfa5)

By using [torch snapshot api ](https://docs.pytorch.org/docs/main/torch_cuda_memory.html#understanding-cuda-memory-usage)to profile, it is found that the GPU memory keeps increasing due to the memory  allocation of multimodal pixel_values [[code link]](https://github.com/sgl-project/sglang/blob/9179ea15957bad2af5f678cf8bbdb5ddc8d2ab41/python/sglang/srt/managers/schedule_batch.py#L1280), The pixel_values ​​tensor is not released in time and keeps occupying GPU memory.
![A7D01595-EB1F-411C-9510-CBE581904EF4_1_201_a](https://github.com/user-attachments/assets/bf3db85c-24c5-4e6d-b05b-9f0702177ad9)

To further verify this, we conducted the following experiment with [code](https://github.com/AniZpZ/sglang/commit/440ffc79297b3ac74a93a74029696674a9d5fb6b)
Record how long it takes for the Req object to be recycled and destroyed after the inference is completed

Intercept part of the log:
```
[2025-06-18 21:18:58 TP0] Prefill batch. #new-seq: 1, #new-token: 1, #cached-token: 15575, token usage: 0.01, #running-req: 0, #queue-req: 0
==================del relay time: 236.99082350730896
==================del relay time: 217.81080794334412
==================del relay time: 213.0320155620575
==================del relay time: 193.54624009132385
==================del relay time: 174.60934591293335
==================del relay time: 169.76573705673218
==================del relay time: 150.77730989456177
==================del relay time: 131.5427701473236
==================del relay time: 126.74609208106995
==================del relay time: 107.52101039886475
==================del relay time: 88.3551561832428
==================del relay time: 83.5151720046997
==================del relay time: 64.09463214874268
==================del relay time: 44.952624797821045
==================del relay time: 40.15912079811096
==================del relay time: 21.200743436813354
==================del relay time: 16.45572304725647
==================del relay time: 11.62731122970581
==================del relay time: 6.783725023269653
==================del relay time: 237.0271179676056
==================del relay time: 217.8470320701599
==================del relay time: 213.06826496124268
==================del relay time: 193.58248448371887
==================del relay time: 174.64558386802673
==================del relay time: 169.80197429656982
==================del relay time: 150.81354665756226
==================del relay time: 131.5790765285492
==================del relay time: 126.77996063232422
==================del relay time: 107.5571517944336
==================del relay time: 88.3913938999176
==================del relay time: 83.55138516426086
==================del relay time: 64.13084888458252
==================del relay time: 44.9889132976532
==================del relay time: 40.19537019729614
==================del relay time: 21.23698902130127
==================del relay time: 16.491973400115967
==================del relay time: 11.663555383682251
==================del relay time: 6.8199708461761475
==================del relay time: 237.0407304763794
==================del relay time: 217.8606629371643
==================del relay time: 213.08187127113342
==================del relay time: 193.5961036682129
==================del relay time: 174.65919375419617
==================del relay time: 169.81558871269226
==================del relay time: 150.8271610736847
==================del relay time: 131.5926854610443
==================del relay time: 126.79595994949341
==================del relay time: 107.57086777687073
==================del relay time: 88.40501689910889
==================del relay time: 83.5650064945221
==================del relay time: 64.14444899559021
==================del relay time: 45.002522468566895
==================del relay time: 40.20898938179016
==================del relay time: 21.25060486793518
==================del relay time: 16.50559091567993
==================del relay time: 11.677167654037476
==================del relay time: 6.8335888385772705
==================del relay time: 237.13922882080078
==================del relay time: 217.95917654037476
==================del relay time: 213.18038606643677
==================del relay time: 193.69461297988892
==================del relay time: 174.757714509964
==================del relay time: 169.9141070842743
==================del relay time: 150.92568612098694
==================del relay time: 131.69119906425476
==================del relay time: 126.8944160938263
==================del relay time: 107.66937947273254
==================del relay time: 88.50351285934448
==================del relay time: 83.66350507736206
==================del relay time: 64.24297046661377
==================del relay time: 45.10103750228882
==================del relay time: 40.307501554489136
==================del relay time: 21.349082469940186
==================del relay time: 16.604097366333008
==================del relay time: 11.775666236877441
==================del relay time: 6.932098388671875
==================del relay time: 237.16592502593994
==================del relay time: 217.9858841896057
==================del relay time: 213.20708107948303
==================del relay time: 193.72132086753845
==================del relay time: 174.78441667556763
==================del relay time: 169.94083333015442
==================del relay time: 150.95237159729004
==================del relay time: 131.71790599822998
==================del relay time: 126.9211630821228
==================del relay time: 107.69609570503235
==================del relay time: 88.53022050857544
==================del relay time: 83.69021701812744
==================del relay time: 64.2696704864502
==================del relay time: 45.12774085998535
==================del relay time: 40.334203004837036
==================del relay time: 21.375828981399536
==================del relay time: 16.630818128585815
==================del relay time: 11.802381992340088
==================del relay time: 6.958810329437256
==================del relay time: 237.17272925376892
==================del relay time: 217.99271154403687
==================del relay time: 213.21391463279724
==================del relay time: 193.72811722755432
==================del relay time: 174.79123187065125
==================del relay time: 169.94764041900635
==================del relay time: 150.95919919013977
==================del relay time: 131.72470903396606
==================del relay time: 126.92797994613647
==================del relay time: 107.70292234420776
==================del relay time: 88.53704738616943
==================del relay time: 83.69703769683838
==================del relay time: 64.27650165557861
==================del relay time: 45.134521484375
==================del relay time: 40.34100532531738
==================del relay time: 21.382639169692993
==================del relay time: 16.637617826461792
==================del relay time: 11.809204816818237
==================del relay time: 6.965626955032349
==================del relay time: 237.18339109420776
==================del relay time: 218.00333905220032
==================del relay time: 213.22454142570496
==================del relay time: 193.73876810073853
==================del relay time: 174.80186915397644
==================del relay time: 169.95826697349548
==================del relay time: 150.96983933448792
==================del relay time: 131.73535132408142
==================del relay time: 126.93862795829773
==================del relay time: 107.71355080604553
==================del relay time: 88.54767894744873
==================del relay time: 83.70766091346741
==================del relay time: 64.28713512420654
==================del relay time: 45.1451940536499
==================del relay time: 40.351662158966064
==================del relay time: 21.39327621459961
==================del relay time: 16.64826273918152
==================del relay time: 11.819843769073486
==================del relay time: 6.976165533065796
[2025-06-18 21:18:59 TP0] Prefill batch. #new-seq: 1, #new-token: 1, #cached-token: 15575, token usage: 0.01, #running-req: 1, #queue-req: 0
[2025-06-18 21:19:00 TP0] Decode batch. #running-req: 2, #token: 15627, token usage: 0.01, cuda graph: True, gen throughput (token/s): 17.78, #queue-req: 0
[2025-06-18 21:19:01 TP0] Decode batch. #running-req: 2, #token: 15707, token usage: 0.01, cuda graph: True, gen throughput (token/s): 122.99, #queue-req: 0
[2025-06-18 21:19:01] INFO:     127.0.0.1:37278 - "POST /v1/chat/completions HTTP/1.1" 200 OK
==================del relay time: 0.0015380382537841797
==================del relay time: 0.0016913414001464844
==================del relay time: 0.0017123222351074219
==================del relay time: 0.0017480850219726562
==================del relay time: 0.001772165298461914
==================del relay time: 0.0017311573028564453
==================del relay time: 0.0018548965454101562
==================del relay time: 0.0018630027770996094
[2025-06-18 21:19:02] INFO:     127.0.0.1:37280 - "POST /v1/chat/completions HTTP/1.1" 200 OK
[2025-06-18 21:19:03 TP0] Prefill batch. #new-seq: 1, #new-token: 1, #cached-token: 15575, token usage: 0.01, #running-req: 0, #queue-req: 0
```
Most Req objects are recycled quickly, but some Req objects have to wait for more than 3 minutes before being recycled, and the pixel_values ​​tensor has been occupying the GPU memory. 8*H20 host has 1.5T cpu memory, so it is estimated that Python's garbage collection is difficult to clean up completely.


Co-author @xu-yfei @yudian0504 
## Modifications
With this fix code, OOM no longer occur.


## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

huangtingwei9988 committed:
fix the abnormal GPU memory occupation of multi-modal models continued to increase until OOM

huangtingwei9988 committed:
fix

Swipe4057 commented:
Please take a look at this issue

https://github.com/huggingface/transformers/issues/38656#issuecomment-2957132567

huangtingwei9988 commented:
> 请查看此问题
> 
> [huggingface/transformers#38656（评论）](https://github.com/huggingface/transformers/issues/38656#issuecomment-2957132567)
> 
> #6933

hi@Swipe4057, We also saw these issue before fixing it, but the main cause of OOM was not the image processor.  The Processor does cause GPU0 to occupy an additional 1.5G of video memory in our test, but that is caused by this [code](https://github.com/sgl-project/sglang/blob/9179ea15957bad2af5f678cf8bbdb5ddc8d2ab41/python/sglang/srt/managers/multimodal_processors/base_processor.py#L128).
 If `kwargs["device"] = "cuda"` is changed to `kwargs["device"] = "cpu"`, the processor memory leak problem will not be on the GPU, but will be transferred to the CPU. However,  the GPU OOM problem still cannot be solved. At the same time, processor will only cause memory leaks in GPU0, but the abnormal memory usage problem I mentioned occurs on all GPUs.

![image](https://github.com/user-attachments/assets/8b474917-2afb-4f39-9e2d-50421fbdb7c3)

yhyang201 commented:
cc @mickqian

huangtingwei9988 committed:
Merge branch 'main' into fix_vl_oom

huangtingwei9988 commented:
> Hi. Have you investigated the reason of this leak? Since normally, the tensor of the `Req` will de deallocated along with `Req` itself.
> 
> And why is it related to `abort`?
> 
> Thanks
hi～@mickqian, As I mentioned above, some Req objects were not released in time after completion. I only ran the experiment for about 20 minutes and found that some Req objects had to wait for more than 3 minutes before being released. As long as the Req object is not released in time, the pixel_values ​​tensor will continue to occupy the GPU memory.

huangtingwei9988 commented:
> Hi, @huangtingwei9988
> 
> I deployed the Qwen2.5-VL-32B model on an 8x H20 GPU setup using the v0.4.7-cu124 docker image, applying your two modified files as you mentioned above. The launch command was:
> 
> `python3 -m sglang.launch_server --model-path /export/Qwen2.5-VL-32B-Instruct --tp 8 --trust-remote-code --chat-template qwen2-vl --host 0.0.0.0`
> 
> However, the GPU memory leak issue persists – each request increases memory usage by ~40MB without being freed. Do you have any suggestions to address this? Any help would be greatly appreciated!

hi~@wiggin66 What you said about each request increases memory usage by ~40MB without being freed, does it happen after sglang is started?  In our test, the GPU memory did increase for a while after it was first started, but then stabilized.
![image](https://github.com/user-attachments/assets/ea6e5e34-aa0c-42c6-8d78-22768eaf2e5c)

yhyang201 commented:
> > > > Hi. Have you investigated the reason of this leak? Since normally, the tensor of the `Req` will de deallocated along with `Req` itself.
> > > > And why is it related to `abort`?
> > > > Thanks
> > > 
> > > 
> > > hi～@mickqian, As I mentioned above, some Req objects were not released in time after completion. I only ran the experiment for about 20 minutes and found that some Req objects had to wait for more than 3 minutes before being released. As long as the Req object is not released in time, the pixel_values ​​tensor will continue to occupy the GPU memory and cause OOM.
> > 
> > 
> > That's what im talking about -- normally the `Req` is expected to deallocate once finished. We'd better find out the reason of this, for a more direct fix.
> 
> Yes! This needs to be further explored. In my opinion, 8*H20 host has 1.5T cpu memory, so it is estimated that Python's garbage collection is difficult to clean up completely.
> 
> I also tried adding the following code to the schduler step, which also prevented OOM
> 
> ```
> import gc
> gc.collect()
> torch.cuda.empty_cache()
> ```

Is this useful? I thought PyTorch automatically reused its cached memory that hasn't been freed back to the OS/driver.

huangtingwei9988 committed:
Merge branch 'main' into fix_vl_oom

huangtingwei9988 committed:
Merge branch 'main' into fix_vl_oom

huangtingwei9988 committed:
Merge branch 'main' into fix_vl_oom

huangtingwei9988 committed:
Merge branch 'main' into fix_vl_oom

huangtingwei9988 committed:
Merge branch 'main' into fix_vl_oom

Swipe4057 commented:
Today I discovered that when enabling --enable-mixed-chunk with high concurrency and request lengths of 2-3 thousand tokens, there is a video memory leak, followed by OOM and reboot. No matter what --mem-fraction-static value I set, OOM still occurs. I tested on 2xH100 and Qwen2.5 32B + JSON output in requests (I haven't yet investigated whether it affects this). This is not an issue with this PR. But since multimodal models can also be run with this setting, the problem will persist, and it improves throughput.

huangtingwei9988 committed:
Merge branch 'main' into fix_vl_oom

huangtingwei9988 commented on python/sglang/srt/managers/schedule_batch.py:
```diff
@@ -646,6 +646,18 @@ def finished(self) -> bool:
         # Whether request reached finished condition
         return self.finished_reason is not None
 
+    def release_mm_resources(self):
```
> Is this all necessary, or `multimodal_inputs = None` suffices?


`del mm_item.pixel_values` this line of code that really matters.

huangtingwei9988 committed:
Merge branch 'main' into fix_vl_oom

ehuaa commented:
> Today I discovered that enabling --enable-mixed-chunk causes video memory leaks when running high concurrency workloads with request lengths of 2-3 thousand tokens. The memory leak leads to OOM errors and system reboots. Regardless of the --mem-fraction-static value I configure, OOM errors still occur. I tested this behavior on 2xH100 with Qwen2.5 32B model using JSON output in requests (I haven't investigated whether JSON output contributes to the memory leak). While the memory leak issue is not related to this specific PR, multimodal models can also use the --enable-mixed-chunk setting, so the problem will affect them as well. The setting does improve throughput performance.

@Swipe4057  Hi, can you share any benchmark settings that can reproduce this OOM error? I have noticed this error before but cannot reproduce it now.

huangtingwei9988 committed:
Merge branch 'main' into fix_vl_oom

huangtingwei9988 committed:
Merge branch 'main' into fix_vl_oom

JustinTong0323 committed:
Merge branch 'main' into fix_vl_oom

JustinTong0323 committed:
Refactor resource release in Req class to improve memory management

Updated the release_mm_resources method to handle multimodal inputs more robustly. Added error handling for attribute access and ensured proper cleanup of resources, specifically removing 'feature' from mm_items. This change enhances memory management and prevents potential errors during resource release.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
Merge branch 'main' into fix_vl_oom

JustinTong0323 committed:
update release mm resources

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
Merge branch 'main' into fix_vl_oom

JustinTong0323 committed:
fix

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

huangtingwei9988 committed:
Merge branch 'main' into fix_vl_oom

handoku commented:
Based on my test on 0.4.9.post6( this pr merged), this pr does not fix anything. The OOM problem persists. One can reproduce easily by serving only `qwen2.5-vl-3b` and having one client sequentially send requests. The GPU memory will increase constantly until OOM.

huangtingwei9988 commented:
> Based on my test on 0.4.9.post6( this pr merged), this pr does not fix anything. The OOM problem persists. One can reproduce easily by serving only `qwen2.5-vl-3b` and having one client sequentially send requests. The GPU memory will increase constantly until OOM.

hi～@handoku, Our solution is effective when our host memory is 1.5TB and 8*H20, and we deploy a large model such as a 72B VLM model. The OOM problem of VLM may be caused by multiple reasons, so our solution may not be effective in your scenario.

handoku commented:
> > Based on my test on 0.4.9.post6( this pr merged), this pr does not fix anything. The OOM problem persists. One can reproduce easily by serving only `qwen2.5-vl-3b` and having one client sequentially send requests. The GPU memory will increase constantly until OOM.
> 
> hi～@handoku, Our solution is effective when our host memory is 1.5TB、 8*H20, and we deploy a large model such as a 72B VLM model. The OOM problem of VLM may be caused by multiple reasons, so our solution may not be effective in your scenario.

Yep, I have opened another issue for a ticket. Thank you for all your tests and contributions. Your fix and test results indeed seem promising, though it didn't fix the problem I encountered. BTY, your code also does no harm, I wish the team would merge your PR ASAP, It may fix a ton of issues.

JustinTong0323 committed:
Merge branch 'main' into fix_vl_oom

JustinTong0323 committed:
Merge branch 'main' into fix_vl_oom

huangtingwei9988 committed:
Merge branch 'main' into fix_vl_oom

huangtingwei9988 committed:
Merge branch 'main' into fix_vl_oom

JustinTong0323 committed:
Merge branch 'main' into fix_vl_oom

JustinTong0323 committed:
Merge branch 'main' into fix_vl_oom

huangtingwei9988 committed:
Merge branch 'main' into fix_vl_oom

huangtingwei9988 committed:
Merge branch 'main' into fix_vl_oom

JustinTong0323 commented:
Also thanks @huangtingwei9988 for the early stage finding of the bug :)
